File: src/18_check_bracket_sequence/check_bracket_sequence.cpp
#include <iostream>
#include <string>

using namespace std;

bool valid(string s) {
  int balance = 0;
  for (auto c : s) {
    if (c == '(') ++balance;
    else --balance;
    
    if (balance < 0) break;
  }
  return balance == 0;
}

int main(void) {
  string s;
  cin >> s;
  string result = valid(s) ? "YES" : "NO";
  cout << result << endl;
  return 0;
}

File: src/18_check_bracket_sequence/check_bracket_sequence.py
def valid(s):
  balance = 0
  for c in s:
    if c == '(':
      balance += 1
    else:
      balance -= 1
      if balance < 0:
        return False
        
  return balance == 0

s = input()
print('YES' if valid(s) else 'NO')
